From 7890b7a6b1fce045520c88542c6df016ae0f372a Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Mon, 28 Aug 2006 20:28:25 +0000
Subject: [PATCH] Fix for CONFIGURATION-222: PropertiesConfiguration.save()
 does not work any more when the configuration was created in memory. Thanks
 to Gabriele Garuglieri for the patch.

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/configuration/trunk@437816 13f79535-47bb-0310-9956-ffa450edef68
---
 .../PropertiesConfiguration.java              | 16 +++++++-
 .../TestPropertiesConfiguration.java          | 39 +++++++++++++++++++
 xdocs/changes.xml                             |  7 ++++
 3 files changed, 61 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/configuration/PropertiesConfiguration.java b/src/java/org/apache/commons/configuration/PropertiesConfiguration.java
index 164cee2441..60598854cf 100644
--- a/src/java/org/apache/commons/configuration/PropertiesConfiguration.java
+++ b/src/java/org/apache/commons/configuration/PropertiesConfiguration.java
@@ -208,6 +208,7 @@ public class PropertiesConfiguration extends AbstractFileConfiguration
      */
     public PropertiesConfiguration()
     {
+        layout = createLayout();
         setIncludesAllowed(false);
     }
 
@@ -340,7 +341,20 @@ public synchronized PropertiesConfigurationLayout getLayout()
      */
     public synchronized void setLayout(PropertiesConfigurationLayout layout)
     {
-        this.layout = layout;
+        // only one layout must exist
+        if (this.layout != null)
+        {
+            removeConfigurationListener(this.layout);
+        }
+
+        if (layout == null)
+        {
+            this.layout = createLayout();
+        }
+        else
+        {
+            this.layout = layout;
+        }
     }
 
     /**
diff --git a/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java b/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java
index 2872db75d6..de3e39bea0 100644
--- a/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestPropertiesConfiguration.java
@@ -158,6 +158,45 @@ public void testSave() throws Exception
         checkConfig.save();
     }
 
+    public void testInMemoryCreatedSave() throws Exception
+    {
+        // remove the file previously saved if necessary
+        if (testSavePropertiesFile.exists())
+        {
+            assertTrue(testSavePropertiesFile.delete());
+        }
+
+        PropertiesConfiguration pc = new PropertiesConfiguration();
+        // add an array of strings to the configuration
+        pc.addProperty("string", "value1");
+        List list = new ArrayList();
+        for (int i = 1; i < 5; i++)
+        {
+            list.add("value" + i);
+        }
+        pc.addProperty("array", list);
+
+        // save the configuration
+        String filename = testSavePropertiesFile.getAbsolutePath();
+        pc.save(filename);
+
+        assertTrue("The saved file doesn't exist", testSavePropertiesFile.exists());
+
+        // read the configuration and compare the properties
+        PropertiesConfiguration checkConfig = new PropertiesConfiguration(filename);
+        for (Iterator i = pc.getKeys(); i.hasNext();)
+        {
+            String key = (String) i.next();
+            assertTrue("The saved configuration doesn't contain the key '" + key + "'",
+                    checkConfig.containsKey(key));
+            assertEquals("Value of the '" + key + "' property", 
+                    pc.getProperty(key), checkConfig.getProperty(key));
+        }
+
+        // Save it again, verifing a save with a filename works.
+        checkConfig.save();
+    }
+
     public void testSaveMissingFilename()
     {
         PropertiesConfiguration pc = new PropertiesConfiguration();
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index a83243b755..8cd264184d 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,13 @@
   <body>
 
     <release version="1.3-rc2" date="in SVN">
+      <action dev="oheger" type="update" issue="CONFIGURATION-222" due-to="Gabriele Garuglieri">
+        The new PropertiesConfigurationLayout class broke the save() operation
+        of PropertiesConfiguration when an instance was newly created and
+        populated in memory. This is fixed now by ensuring that a layout object
+        is immediately created and registered as event listener at the
+        configuration.
+      </action>
       <action dev="oheger" type="add" issue="CONFIGURATION-221" due-to="Rainer Jung">
         ConfigurationFactory now supports variables in its configuration
         definition files. These variables are resolved using system properties
